### DISCUSSION TITLE
How does keycloak works in a microservice system with oauth2 ?

### DISCUSSION THREAD

**Discussion by Tockra** [keycloak/keycloak#15492] [Q&A]
Hi,

I'm looking for a tutorial / guide, which explains how to use keycloak in a microservice architecture with oauth2 (frontend single page application + microservice backend). 
I'm pretty sure that there is a possibility to validate an token and get the Roles from the token without talking to keycloak. 
The question is, if this is recommended. Or should every microservice get an keycloak-client with credential and check the roles through an request to keycloak?
Further how/when/where should the whole system check the valid state of the users token?

I looked for documentation about this topic, but I didn't found something helpful :(

T

**Comment by lexcao** [comment]
Is there a gateway in your microservice backend?
It is better to validate a token and get roles from the gateway, 
then pass verified user info and roles to the microservice backend via additional request headers.

**Comment by Tockra** [comment]
Yes, there is a java gateway. But we plan to replace it in future, because we think a nginx gateway should be more performant. 
About what kind of validation do you speak? A local validation on the gateway or a request to keycloak which checks the validity of the token?

**Reply by lexcao** [reply]
Request to the Keycloak to get the public JWS info. 
* Could cache, then update with interval
Then verify the token locally with the public key on the gateway.

**Comment by embesozzi** [comment]
HI @Tockra 
The recommendation is to follow the typical authorization architecture which have the following components:
1. Authorization Server: Responsible for authenticating the users and issuing tokens. Component: Keycloak
2. Policy Enforcement Point (PEP): intercepts a user’s access to a resource, and either grants or denies access to the resource requested by the user. PEPs don't make the decisions; they enforce them. For Coarse-Grained Authorization (CGA) you could use an API Gateway and for Fine-Grained authorization (FGA) you could apply the enforcement on API Side.
3. Policy Decision Point (PDP): evaluates authorization requests against relevant policies and makes an authorization decision. Here you have access control rules in business logic for instance on the API Side or you can follow the approach of decoupling the authorization policies from the code with [OPA](https://www.openpolicyagent.org/docs/latest/).
4. Policy Information Point (PIP): any data source that contains attributes that are relevant for making policy decisions. 

Based on your initial question, you could add the KC roles into the JWT and then the API Gateway or API will validate the token and extract the claim "roles" for applying the access control rules. There is a limit in the HTTP Header size, so, try not to put too much information inside the JWT.

**Comment by Tockra** [comment]
About 2. I have two question:
1. Can this PEP only be a java application or is there a configuration for nginx to use this application?
2. Isn't it a bottle neck to start validation of an token before passing the request to the next point in the cluster?

If I understand you correct, there will be never a check against keycloak and only be client site validation. Or when is it not recommended to do it locally. 
If I use RBAC and give a user a role in keycloak, then the old users token doesn't contain the new role until the token is expired. But this isn't nice.

**Reply by embesozzi** [reply]
My answers below:

1. It's depend on your requirements and the authz architecture that you want to implement. You can implement the PEP in your java application. You have an example in the following project: [embesozzi/oauth-demoapi-jwt-rbac-spring](https://github.com/embesozzi/oauth-demoapi-jwt-rbac-spring). This is a Spring Boot app acting as OAuth2 Resource Server applying the enforcement based on the kc roles inside the JWT.
On the other hand, you can also implement the PEP in the API Gateway, for instance, Kong or Kong + OPA, etc

2. Every API protected by OAuth is doing that. The API is just validating the JWT and then applying the access control rules based on identity claims.

> If I understand you correct, there will be never a check against keycloak and only be client site validation. 

Yes, because the token is a JWT and the approach that you are following is protecting an API with OAuth2.

> If I use RBAC and give a user a role in keycloak, then the old users token doesn't contain the new role until the token is expired. 

Each approach has its pros and cons, life is hard :smile:. This is why is critical to know in advance how critical the app / api is in order to define which is the best approach to follow.

**Reply by Tockra** [reply]
Okay, thank you for your answeres. But won't be there a bottle neck in the whole cluster (in PEP), if there will be (often) be a request to the keycloak instance to get all roles and build them into a jwt token ?
Wouldn't be it better for load balance of the cluster to distribute the different keycloak request to the specific applications, which needs the keycloak role for authorization ?

**Reply by embesozzi** [reply]
As I was saying at the beginning, it's impossible to answer those questions in advance without having requirements, number of identities, roles, integrated apps, architecture design, performance tests, etc.  You have differents approach in the authz architecture scenario, each one has their pro and cons.
Based on your last question, you could have N instances running in K8s for handling the authn + authz processes, you could add the roles to the token (AT) during the authentication process or for some particular use cases you could add the roles only when the app is calling the userinfo endpoint. But you control the information that you want to add in the token with OAuth scopes.

My recommendation is there is no fixed path to follow, you can adapt the API protection guidelines based on your requirements. Nevertheless, I recommend you to follow the standards like OAuth 2.0, UMA, etc.
